haskell/test/Model/BaseDAO/Proposal/FreezeHistory.hs

```diff
@@ -0,0 +1,54 @@
+-- SPDX-FileCopyrightText: 2021 TQ Tezos
+-- SPDX-License-Identifier: LicenseRef-MIT-TQ
+
+module Model.BaseDAO.Proposal.FreezeHistory
+  ( addFrozenFh
+  , subFrozenFh
+
+  , stakeFrozenFh
+  , unstakeFrozenFh
+
+  , updateFh
+  ) where
+
+import Universum
+
+import Control.Monad.Except (throwError)
+import GHC.Natural (minusNaturalMaybe)
+
+import Ligo.BaseDAO.Types
+import Model.BaseDAO.Types
+
+addFrozenFh :: Natural -> AddressFreezeHistory -> AddressFreezeHistory
+addFrozenFh amt fh = fh { fhCurrentUnstaked = (fh & fhCurrentUnstaked) + amt}
+
+subFrozenFh :: Natural -> AddressFreezeHistory -> ModelT AddressFreezeHistory
+subFrozenFh amt fh =
+  case (fh & fhPastUnstaked) `minusNaturalMaybe` amt of
+    Just newAmt -> pure $ fh { fhPastUnstaked = newAmt }
+    Nothing -> throwError NOT_ENOUGH_FROZEN_TOKENS
+
+stakeFrozenFh :: Natural -> AddressFreezeHistory -> ModelT AddressFreezeHistory
+stakeFrozenFh amt fh = do
+  updatedFh <- subFrozenFh amt fh
+  pure (updatedFh { fhStaked = (updatedFh & fhStaked) + amt })
+
+unstakeFrozenFh :: Natural -> Natural -> AddressFreezeHistory -> ModelT AddressFreezeHistory
+unstakeFrozenFh amtToUnstake amtToBurn fh =
+  case (fh & fhStaked) `minusNaturalMaybe` (amtToUnstake + amtToBurn) of
+    Just newStakedAmt ->
+      pure $ fh { fhStaked = newStakedAmt, fhPastUnstaked = (fh & fhPastUnstaked) + amtToUnstake }
+    Nothing ->
+      error "BAD_STATE: Unstake amount is more than staked amount."
+
+updateFh :: Natural -> AddressFreezeHistory -> AddressFreezeHistory
+updateFh currentStage freezeHistory =
+  if (freezeHistory & fhCurrentStageNum) < currentStage then
+    AddressFreezeHistory
+      { fhCurrentStageNum = currentStage
+      , fhStaked = freezeHistory & fhStaked
+      , fhCurrentUnstaked = 0
+      , fhPastUnstaked = (freezeHistory & fhCurrentUnstaked) + (freezeHistory & fhPastUnstaked)
+      }
+  else freezeHistory
```

Comment:
Here's my suggestions to move this away from the LIGO source.

These should be only 3 "updating" functions:
- one for freezing
- one for staking
- one for burning

The one for freezing should take an `Integer` to add or subtract to current/past unstaked (respectively)

The one for staking should take an `Integer` to stake/unstake tokens (and move to/from the unstaked fields).

The one for burning should only burn from staked tokens.

All 3 should:
1. update the freeze history before doing anything else (`updateFh` should not be exported IMO)
2. take the owner of these freeze history and fetch/update the value from/to the storage on their own

This would bring IMO enough variance to make these tests meaningful.
Also please comment on each of these functions describing their behavior.

Reply:
I've added this 3 updates function now, let me know if this is the same as what you have in mind.